### ISSUE TITLE
Feature: Built-in AutoEq

### ISSUE THREAD

**Issue by bernharl** [wwmm/easyeffects#2027]
Hi,

I think it would be great for for Easyeffects to have the FIR filters from https://github.com/jaakkopasanen/AutoEq built in to the convolver plugin. How desirable / feasible is this? This would get headphone tuning in Easyeffects closer to the ease of use provided by Wavelet https://pittvandewitt.github.io/Wavelet/ .

Is this a feasible addition to Easyeffects? :)

**Comment by pinkflames** [comment]
I believe the AutoEQ dataset comes from various published papers on HRTF. I'm no expert on this but just because something has been published in a research paper does not mean it's void of intellectual property such s copyrights and patents, so, while it might be enough to just *cite the original source of each profile*, distributing such dataset and using it in 3rd party software might be uncharted territory.

**Comment by malinges** [comment]
FWIW it looks like AutoEq presets come from https://github.com/jaakkopasanen/AutoEq and are [MIT licensed](https://github.com/jaakkopasanen/AutoEq/blob/master/LICENSE).

**Comment by pinkflames** [comment]
Exactly. I could be wrong, since this is not my area of expertise but I believe that the MIT license does not require assigning patent use rights to source code recipients and is therefore compatible with publishing code that's actually encumbered and not free to distribute. While possibly dubious it's likely that MIT license could be applied to copyrighted material without breaking its licensing, as well. As I said, this is likely a legal grey area but I'm not a lawyer.

LIkewise, I think, this was discussed over at PipeWire project not that long ago and the conclusion was that, while we could add the ability to use AutoEq profiles, if present, most likely no legally competent distro would dare to ship the AutoEq dataset due to its dubious legal standing. But this is my personal opinion and is by no means legal advice nor is this the opinion of the PipeWire project.

**Comment by wwmm** [comment]
Independently of the possible legal issues I don't think that embedding more than 2500 impulse files in our sources and making sure they are updated is the way to go.

**Comment by magicgoose** [comment]
IMO as a user, copying an impulse file is not that hard.
However, it would be convenient of course, if it auto-detected the connected bluetooth headphones and downloaded the correct .wav files from AutoEQ git repository. But it's also more complicated than it sounds. For many headphones, there are different curves for when they are in active or passive mode, etc.
As long as the manual routine of importing a .wav impulse is kept straightforward, I don't really see the need.

**Comment by wabkia** [comment]
no reason not to have it. peace APO has had it for years

**Comment by wwmm** [comment]
>doesnt need to do anything other than have a button to apply autoeq presets.

Let's be reasonable. A button does nothing by itself. So it is more than just putting a button in the window. The impulse response files have to come from somewhere. And we have to change our window in a way that the user can import from a list of AutoEQ files just what is needed.

How is peace apo handling the problem? As long as it does not involve copying thousands of impulse files to EasyEffects source code folder I am willing to listen.

**Comment by edsimpsons83** [comment]
> How is peace apo handling the problem? As long as it does not involve copying thousands of impulse files to EasyEffects source code folder I am willing to listen.

On Windows, clicking the AutoEQ button pops up a window where you can search for the preset you want and then hit "use" to apply it. This seems similar to how SoundSource and eqMac built it in on OSX.

![AUTOEQ](https://github.com/wwmm/easyeffects/assets/27150464/3971f31e-6639-4dec-be82-abadd24d4a80)

**Comment by Digitalone1** [comment]
Now it's the time someone should make a community preset with all jaakkopasanen filters.

**Comment by wwmm** [comment]
> > How is peace apo handling the problem? As long as it does not involve copying thousands of impulse files to EasyEffects source code folder I am willing to listen.
> 
> On Windows, clicking the AutoEQ button pops up a window where you can search for the preset you want and then hit "use" to apply it. This seems similar to how SoundSource and eqMac built it in on OSX.
> 
> ![AUTOEQ](https://private-user-images.githubusercontent.com/27150464/325315913-3971f31e-6639-4dec-be82-abadd24d4a80.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM5OTY5NjYsIm5iZiI6MTcxMzk5NjY2NiwicGF0aCI6Ii8yNzE1MDQ2NC8zMjUzMTU5MTMtMzk3MWYzMWUtNjYzOS00ZGVjLWJlODItYWJhZGQyNGQ0YTgwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDI0VDIyMTEwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTljYjk4ZmQyYWVhN2RmNDViM2E4ZDIxMDIyOWZlMGExM2U2N2VjYWQwZTYyZGExMmY0YjI0YzRiZjk1YWM5Y2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.sirEpDKkgk_oUGIdmhOo7F4NAwt7STtllaaj5kortEQ)

Ok. But the main problem is the location of all these impulse files. Are they downloading the file list directly from the AutoEq site and then download only the file selected by the user? Or is this list coming from files shipped together with AutoEQ? Having thousands of impulse files suddenly copied to our sources is definitely not happening. The other approach will depend on how hard it is to do...

>Now it's the time someone should make a community preset with all jaakkopasanen filters.

It would definitely be the easier solution. And would use all that infrastructure that was put in place for community presets.

**Comment by edsimpsons83** [comment]
> Are they downloading the file list directly from the AutoEq site and then download only the file selected by the user? Or is this list coming from files shipped together with AutoEQ? Having thousands of impulse files suddenly copied to our sources is definitely not happening. The other approach will depend on how hard it is to do...

It's the former. The window seems to grab a list of all the AutoEq options. When you pick one to use, it downloads the file and applies the settings.

**Comment by Digitalone1** [comment]
The problem with those AutoEQ is that they are too many. They are also in APO which could only be parsed by the app inserting them manually in the Equalizer UI.

Someone with very high perseverance can make a script to convert them all in our json format. You should start from [this](https://github.com/wwmm/easyeffects/blob/master/src/equalizer_ui.cpp#L358-L446) and adapt the code to make a json preset.

It's not easy. Maybe in Javascript is good idea. Good luck.

**Comment by korpe1** [comment]
I think, we should always download the latest response files directly from AutoEq, instead of shipping the dataset in the EasyEffects package. And then we come the problem of rely one online service. And because of that, I suggest implement this in a plugin. Please dont't bolt this down to EasyEffects.

The plugin would fetch the latest files from EasyEQ with something "standardized" API. I don't mean make a real standard, but sort of _de-facto_ solution. And by that allow users to ability to add also other sources than just EasyEQ.
Cooperation with EasyEQ is crucial in this because it's so important and a big response file dataset.

What do you think @wwmm?

CC: @jaakkopasanen

**Comment by Digitalone1** [comment]
@korpe1  That won't happen.

With all the effort we made for community presets, which was hard to implement, but still easier than fetching things online, someone should make a script to convert all the AutoEQ APO profiles to EE presets and run it regularly to keep it up to date.

Meanwhile a community package is published on Flatpak so you can install it and use inside EE.

**Comment by korpe1** [comment]
> Meanwhile a community package is published on Flatpak so you can install it and use inside EE.

What is a community package?

**Comment by korpe1** [comment]
> AutoEQ APO profiles to EE presets

Have EasyEffects own format?

AFAIK EasyEffects support only preset formats "APO" and "GraphicEQ". Btw is it right that these both are formats from the Equalizer APO application?

**Comment by Digitalone1** [comment]
> > Meanwhile a community package is published on Flatpak so you can install it and use inside EE.
> 
> What is a community package?

A new feature of the next release:

https://github.com/wwmm/easyeffects/blob/master/COMMUNITY_PRESETS_GUIDELINES.md

**Comment by Digitalone1** [comment]
> Have EasyEffects own format?
> 
> AFAIK EasyEffects support only preset formats "APO" and "GraphicEQ". Btw is it right that these both are formats from the Equalizer APO application?

Yes, EE has it's own format. The Equalizer has two functions for converting APO and GraphicEQ, but the community presets should have EE format to be loaded.

**Comment by korpe1** [comment]
> Yes, EE has it's own format. The Equalizer has two functions for converting APO and GraphicEQ, but the community presets should have EE format to be loaded.

What is the need to use a new format for presets?

Does anyone know this:
"AFAIK EasyEffects support only preset formats "APO" and "GraphicEQ". Btw is it right that these both are formats from the Equalizer APO application?"

**Comment by wwmm** [comment]
>What is the need to use a new format for presets?

@korpe1 EasyEffects does not manage only the equalizer plugin. There are several others there. When we talk about preset format we mean the json file where the settings of all plugins can be saved. The APO support in the EQ window is just a convenience to help people import equalization settings from APO files. Nothing more than that.

When I started this project support for APO presets did not even exist because the equalizer I used at the time did not offer an "APO friendly mode". There was never a reason to make the APO standard a first client citizen in EasyEffects. And probably there will never be.

**Comment by wwmm** [comment]
>Are APO and GraphicEQ files from the Equalizer APO application? That application for Windows if I remember right?

Yes. AutoEQ uses the format from the Windows APO EQ.

**Comment by fhuehfig** [comment]
JamesDSP (JDSP) have this function long time ago
https://flathub.org/ru/apps/me.timschneeberger.jdsp4linux
https://github.com/Audio4Linux/JDSP4Linux
And AutoEQ site add support for Easy Effect.
- - -
<img width="1140" height="120" alt="Image" src="https://github.com/user-attachments/assets/836f921c-f228-4734-be1e-aecfac71829e" />
